rfd/0001-testing-guidelines.md

```diff
@@ -0,0 +1,207 @@
+with every new Go release. We just need to complement it with:
+- asserts
+- flexible value diffing (got vs want)
+
+## Details
+
+Use the `testing` package as a base and complement it with several libraries.
+
+### Test organization
+
+Use a hierarchy of:
+- `_test.go` files
+- `func Test*` functions
+- [`t.Run`](https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks)
+  subtests (aka table-driven tests)
+
+Shared setup/teardown:
+- [`func TestMain`](https://golang.org/pkg/testing/#hdr-Main) for package level
+- a single `func Test*` with subtests for smaller test groups
+- `func setupX` called from multiple `func Test*` for larger test groups
+
+For performance testing use
+[benchmarks](https://golang.org/pkg/testing/#hdr-Benchmarks).
+
+### Parallelization
+
+Call [`t.Parallel()`](https://golang.org/pkg/testing/#T.Parallel) from all
+tests and subtests that don't need to run serially.
+
+### Asserts
+
+Use
+[testify/assert](https://pkg.go.dev/github.com/stretchr/testify/assert?tab=doc)
+or a plain `if` condition, whichever is easiest.
+
+Skim through the `testify/assert` docs, it has many convenient helpers.
+
+### Diffing
+
+For trivial comparisons, use helpers from
+[testify/assert](https://pkg.go.dev/github.com/stretchr/testify/assert?tab=doc).
+
+For non-trivial comparisons (such as deeply-nested structs or protobufs), use
+[go-cmp](https://pkg.go.dev/github.com/google/go-cmp/cmp?tab=doc) with
+[cmpopts](https://pkg.go.dev/github.com/google/go-cmp@v0.5.1/cmp/cmpopts?tab=doc).
+
+It allows you to customize checking by ignoring types/fields, equating empty
+and nil slices, approximating float comparisons. See examples below.
+
+### Wishlist
+
+Some things we don't have a good solution for yet:
+- log output collection
+  - if something is logged by a function under test and that test fails, print
+    the log output
+  - it's currently hard to separate logging from a function under test and
+    unrelated logs by a background goroutine
+- performance regression tracking
+  - `go test` will report runtimes of individual tests/subtests
+  - but we don't have tooling to track those over time
+  - note: [benchmarks](https://golang.org/pkg/testing/#hdr-Benchmarks) are an
+    alternative, but must be written manually
+
+## Migration
+
+Re-writing all of teleport tests is a large task. To make it feasible,
+migration will happen organically over the course of ~1 year after these
+guidelines are approved.
+
+The approach is:
+1. no unit tests are rewritten upfront
+1. any new test should follow these guidelines
+1. updates to any existing test (as part of PRs) should be bundled with a
+  migration of that test
+
+There are a few exceptions to #1:
+- `/integration` tests
+- `/lib/services/...` tests
+
+These are massive test suites that are ripe for refactoring. Rewriting them
+will help us flesh out any unforeseen issues with the new testing guidelines.
+
+If after 1 year we're left with un-migrated tests, we will discuss whether it's
```

Comment:
Looks good. Do we want to add checkpoints to review every quarter so we can track progress?

Reply:
( Send an invite for 2021 )

Reply:
It would be nice to track progress.
But I also don't think we'd change our behavior (migrate when updating relevant tests) based on the numbers.
The idea here is "don't worry about migration until it's taking too long".

I'll make myself a small script for tracking, but let's not have multiple people spending time on this yet.